### PR TITLE
Implement capability for creating user-defined isotopes

### LINKED ISSUE #145: Add a new `CustomIsotope` class.
```py
class CustomIsotope:
   symbol: str
   spin: int
   natural_abundance: float
   gyromagnetic_ratio: float
   quadrupole_moment: float
   atomic_number: int
```

- [ ] Check serialization and reading custom isotopes.

### PR DESCRIPTION
**Pull request by mgiammar**
Closes #145 

- [x] Write custom isotope function
- [x] Test new code
- [x] Document new features (changelog & api)

### CODE REVIEW

File: src/mrsimulator/spin_system/isotope.py [hunk 1]
```diff
@@ -103,6 +116,64 @@ def larmor_freq(self, B0=9.4):
...
 
 
+def add_custom_isotope(
+    symbol: str,
+    spin: float,
+    gyromagnetic_ratio: float,
+    quadrupole_moment: float = 0,
+    natural_abundance: float = 100,
+    atomic_number: int = -1,  # What to put for this value??
+):
```
Comment (deepanshs) [inline src/mrsimulator/spin_system/isotope.py:126]: Can this be a classmethod of `Isotope` class? Something like, `Isotope.add_new()`!

File: src/mrsimulator/spin_system/isotope.py [hunk 2]
```diff
@@ -49,10 +50,86 @@ class Config:
...
+        """
+        # Check for symbol overlap in dictionaries
+        if symbol in ISOTOPE_DATA:
+            raise ValueError(
+                "Custom isotope symbol cannot match a real isotope symbol. "
+                f"Got {symbol}."
+            )
+
+        if symbol in custom_isotope_data:
+            return Isotope(symbol=symbol)
```
Comment (deepanshs) [inline src/mrsimulator/spin_system/isotope.py:98]: If the symbol already exits in `custom_isotope_data`, it should raise an error. Consider two entries with same symbol but different spin. At present, the first entry will take priority. The better approach is to not allow duplicate symbols.

Comment (deepanshs) [reply src/mrsimulator/spin_system/isotope.py:98]: Remove atomic number from the custom isotopes. While real isotopes share symbol but vary in atomic number, custom isotropic should be unique symbols.

Just return `0` from `def atomic_number` when attribute is no present.

File: src/mrsimulator/spin_system/isotope.py [hunk 3]
```diff
@@ -49,10 +50,86 @@ class Config:
...
 
     @validator("symbol", always=True)
-    def get_isotope(cls, v, *, values, **kwargs):
+    def validate_symbol(cls, v, *, values, **kwargs):
+        if v in custom_isotope_data:
+            return v
         return format_isotope_string(v)
 
+    @classmethod
+    def add_new(
```
Comment (deepanshs) [inline src/mrsimulator/spin_system/isotope.py:59]: How about the function name `def new()`?

**Review by deepanshs — CHANGES_REQUESTED** [review]

**Comment by deepanshs** [comment]
How about a utility function for the Isotope class that creates a copy of the in-build isotopes as a custom isotope but with a different symbol?

```py
Ha = Isotope.copy(symbol='Ha', from='1H')
```

I see this being a lot useful.

**Comment by deepanshs** [comment]
> How about a utility function for the Isotope class that creates a copy of the in-build isotopes as a custom isotope but with a different symbol?
> 
> ```python
> Ha = Isotope.copy(symbol='Ha', from='1H')
> ```
> 
> I see this being a lot useful.

On second thought, 

```py
Ha = Isotope.new(symbol='Ha', copy_from='1H')
```
might be better. A single function `new` will generate a custom isotope, and if `copy_from` is provided, a custom isotope from the given `copy_from` isotope.

**Comment by mgiammar** [comment]
> > How about a utility function for the Isotope class that creates a copy of the in-build isotopes as a custom isotope but with a different symbol?
> > ```python
> > Ha = Isotope.copy(symbol='Ha', from='1H')
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I see this being a lot useful.
> 
> On second thought,
> 
> ```python
> Ha = Isotope.new(symbol='Ha', copy_from='1H')
> ```
> 
> might be better. A single function `new` will generate a custom isotope, and if `copy_from` is provided, a custom isotope from the given `copy_from` isotope.

I agree a function which copies an isotope but with a new symbol would be the most useful; this is easily implemented with a mapping where the new symbol points to the original symbol and the original symbol points to the `isotope_data.json` data. This way no new information needs to be stored and the mapping can be serialized with the `spin_systems` list.

As we discussed in meeting, I don't think non-real isotopes would be that useful

**Comment by deepanshs** [comment]
Here are some errors I get when running code through this PR.

- In a notebook env, if I run this following code block once, it fine. 

```py
a = Isotope.new(symbol='H1a', copy_from='1H')
```

If I rerun, it gives the following error

```shell
mrsimulator/mrsimulator/src/mrsimulator/spin_system/isotope.py in new(cls, symbol, copy_from)
     77         """
     78         if symbol in ISOTOPE_DATA or symbol in custom_isotope_mapping:
---> 79             raise ValueError(
     80                 "The new symbol for the copied isotope cannot match any previous "
     81                 f"isotope symbols. Got {symbol} which matches a known symbol."

ValueError: The new symbol for the copied isotope cannot match any previous isotope symbols. Got H1a which matches a known symbol.
```

**Comment by deepanshs** [comment]
- Running the simulation with custom isotope.

```py
a = Isotope.new(symbol='H1a', copy_from='1H')
sys = SpinSystem(sites=[Site(isotope='H1a')])
method = BlochDecaySpectrum(channels=['H1a'])
sim = Simulator(spin_systems=[sys], methods=[method])
sim.run()
```

gives a value error. The error log is long. It seems to originate from the parallel run.
 
```shell
ValueError: need more than 1 value to unpack
```

**Comment by deepanshs** [comment]
- Reading spin systems from file without the `isotope_mappings` key,

```shell
File "/home/docs/checkouts/readthedocs.org/user_builds/mrsimulator/checkouts/226/examples_source/2D_simulation(crystalline)/plot_6_PASS_itraconazole_drug.py", line 30, in <module>
    sim.load_spin_systems(filename)
  File "/home/docs/checkouts/readthedocs.org/user_builds/mrsimulator/envs/226/lib/python3.8/site-packages/mrsimulator-0.8.0.dev0-py3.8-linux-x86_64.egg/mrsimulator/simulator/__init__.py", line 274, in load_spin_systems
    _ = [Isotope.new(k, v) for k, v in contents["isotope_mappings"].items()]
TypeError: list indices must be integers or slices, not str
```

File: src/mrsimulator/spin_system/isotope.py [hunk 4]
```diff
@@ -49,10 +51,47 @@ class Config:
...
+        if symbol in ISOTOPE_DATA or symbol in custom_isotope_mapping:
+            raise ValueError(
+                "The new symbol for the copied isotope cannot match any previous "
+                f"isotope symbols. Got {symbol} which matches a known symbol."
+            )
+
+        # Check if there is symbol to map
+        if copy_from not in ISOTOPE_DATA:
+            raise ValueError(
+                "The copy_from symbol must match a known isotope, but "
```
Comment (deepanshs) [inline src/mrsimulator/spin_system/isotope.py:87]: maybe just have this message. 

```py
f"Unknown isotope symbol: {copy_from}"
```

**Comment by mgiammar** [comment]
New features moved to #250

**Pull request closed without merging**